### DISCUSSION TITLE
Can Apiman support other protocols/patterns than HTTP+REST (e.g. Websockets, messaging, event sourcing, etc)

### DISCUSSION THREAD

**Discussion by jintaosai314** [apiman/apiman#1116] [Ideas]
When can apiman support services such as eventsource or websocket, not just content type like JSON or XML.

**Comment by msavy** [comment]
At the moment, no. We only really support HTTP+REST (or SOAP over HTTP in a very simple way). 

But this is a great question and an area that I've been interested in exploring for several years. I'd love to get more community input about what you want to do with websockets/event sourcing/messaging, so that we can figure out the best way to go about this.

So, everyone, let us known your use cases for non-HTTP things, and how you would like Apiman to behave :-)!

p.s. I love eventsourcing, so would love a way to figure out how to support that, either explicitly or implicitly.

**Reply by jintaosai314** [reply]
Thank you for your reply. I think we may encounter such scenarios where server sent events need to be used, such as pushing the real-time location of a taxi through the server. This requires apiman to support the event stream type.

**Reply by msavy** [reply]
That's with WebSockets is it?

And what sort of policies do you use?

**Reply by jintaosai314** [reply]
![diff](https://user-images.githubusercontent.com/62642433/116024452-cec56600-a680-11eb-9af2-8a78626d512c.PNG)
EventSource is a less-powerful way of communicating with the server than WebSocket.

Why should one ever use it?

The main reason: it’s simpler. In many applications, the power of WebSocket is a little bit too much.

We need to receive a stream of data from server: maybe chat messages or market prices, or whatever. That’s what EventSource is good at. Also it supports auto-reconnect, something we need to implement manually with WebSocket. Besides, it’s a plain old HTTP, not a new protocol.

**Reply by msavy** [reply]
Ah OK, you are talking specifically about EventSource not [event sourcing](https://microservices.io/patterns/data/event-sourcing.html). I misunderstood you.

Supporting WebSockets and other message-y things is certainly something that is interesting to us, if we can think of how to support it cleanly.
